rllib/env/base_env.py

```diff
@@ -245,10 +258,72 @@ def action_space(self) -> gym.Space:
         """
         raise NotImplementedError
 
+    @PublicAPI
+    def action_space_sample(self, agent_id: list = None) -> MultiEnvDict:
+        """Returns a random action for each environment, and potentially each
+            agent in that environment.
+
+        Args:
+            agent_id: List of agent ids to sample actions for. If None or empty
+                list, sample actions for all agents in the environment.
+
+        Returns:
+            A random action for each environment.
+        """
+        del agent_id
+        return {}
+
+    @PublicAPI
+    def observation_space_sample(self, agent_id: list = None) -> MultiEnvDict:
+        """Returns a random observation for each environment, and potentially
+            each agent in that environment.
+
+        Args:
+            agent_id: List of agent ids to sample actions for. If None or empty
+                list, sample actions for all agents in the environment.
+
+        Returns:
+            A random action for each environment.
+        """
+        logger.warning("observation_space_sample() has not been implemented")
+        return {}
+
+    @PublicAPI
+    def last(self) -> Tuple[MultiEnvDict, MultiEnvDict, MultiEnvDict,
+                            MultiEnvDict, MultiEnvDict]:
+        """Returns the last observations, rewards, and done flags that were
+            returned by the environment.
+
+        Returns:
+            The last observations, rewards, and done flags for each environment
+        """
+        logger.warning("last has not been implemented for this environment.")
+        return {}, {}, {}, {}, {}
+
+    @PublicAPI
     def observation_space_contains(self, x: MultiEnvDict) -> bool:
```

Comment:
wait, is the type still MultiEnvDict here and below??
I thought we are saying gym and multi-agent envs return different types now?

Reply:
Yes -- the return type of poll() and try_reset() are MultiEnvDicts, and so I thought it would be appropriate if observations and actions produced by the environment/policy should be able to be easily passed to the environment for checking. 

We could probably add functions here for sampling from the observation space, that produce MultiEnv dicts...actually we likely will have to -> there isn't a way to know how to construct this MultiEnvDict for actions and obs, because some environments will have agents, and others won't. Some strange logic will have to be used to infer how to build MultiEnvDicts in the Multi-agent case, without knowing that the environment is multi-agent. 

Thanks for raising this concern. It revealed that there is an inconsistency that I proposed I'm going to add these functions for sampling from the spaces.

Reply:
I might add it in a later PR.

Reply:
ok, it seems to me we are settling on using multi-agent api for single agent env as well, which is totally fine, and probably logical.

Reply:
The BaseEnv is never a single-agent env. If there is only one agent and we derive the BaseEnv from e.g. a gym.Env, we auto-create "DUMMY_AGENT_ID" in the env as the agent's ID.